clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +348,560 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+            revocationResult = CompletableFuture.completedFuture(null);
+            // Reschedule the auto commit starting from now (new assignment received without any
+            // revocation).
+            commitRequestManager.resetAutoCommitTimer();
+        }
+
+        // Future that will complete when the full reconciliation process completes (revocation
+        // and assignment, executed sequentially)
+        CompletableFuture<Void> reconciliationResult =
+                revocationResult.thenCompose(__ -> {
+                    boolean memberHasRejoined = !Objects.equals(memberIdOnReconciliationStart,
+                            memberId);
+                    if (state == MemberState.RECONCILING && !memberHasRejoined) {
+
+                        // Apply assignment
+                        CompletableFuture<Void> assignResult = assignPartitions(assignedPartitions, addedPartitions);
+
+                        // Clear topic names cache only for topics that are not in the subscription anymore
+                        for (TopicPartition tp : revokedPartitions) {
+                            if (!subscriptions.subscription().contains(tp.topic())) {
+                                assignedTopicNamesCache.values().remove(tp.topic());
+                            }
+                        }
+
+                        return assignResult;
+
+                    } else {
+                        String reason;
+                        if (state != MemberState.RECONCILING) {
+                            reason = "The member already transitioned out of the reconciling " +
+                                    "state into " + state;
+                        } else {
+                            reason = "The member has re-joined the group.";
+                        }
+                        // Revocation callback completed but the reconciled assignment should not
+                        // be applied (not relevant anymore). This could be because the member
+                        // is not in the RECONCILING state anymore (fenced, failed, unsubscribed),
+                        // or because it has already re-joined the group.
+                        CompletableFuture<Void> res = new CompletableFuture<>();
+                        res.completeExceptionally(new KafkaException("Interrupting reconciliation" +
+                                " after revocation. " + reason));
+                        return res;
+                    }
+                });
+
+        reconciliationResult.whenComplete((result, error) -> {
+            markReconciliationCompleted();
+            if (error != null) {
+                // Leaving member in RECONCILING state after callbacks fail. The member
+                // won't send the ack, and the expectation is that the broker will kick the
+                // member out of the group after the rebalance timeout expires, leading to a
+                // RECONCILING -> FENCED transition.
```

Comment:
What happen in this case? I suppose that the reconciliation will be retried. Did I get it right?

Reply:
Yes, it will be retried on the next reconciliation loop (known shortcoming how we trigger the reconciliation loops, will be improved right after with [KAFKA-15832](https://issues.apache.org/jira/browse/KAFKA-15832)